File: arrangeCoins/arrangeCoins.php
<?php
/**
 * Created by PhpStorm.
 * User: Hhx
 * Date: 2019/12/23
 * Time: 14:46
 */
class Solution {

    /**
     * @param Integer $n
     * @return Integer
     */
    function arrangeCoins($n) {
        return (int) (sqrt(2 * $n + 1/4) - 0.5);

        // $left = 1;
        // $right = $n;
        // while($left <= $right){
        //     $mid =(int)($left + ($right - $left) / 2) ;
        //     $sum = $mid * ($mid + 1) / 2;
        //     if ($sum == $n) {
        //         return $mid;
        //     }else if ($sum > $n) {
        //         $right = $mid - 1;
        //     }else {
        //         $left = $mid + 1;
        //     }
        // }
        // return $right;
    }
}

File: arrangeCoins/arrangeCoins.go
/*
@Time : 2019/12/23 15:40
@Author : hhx06
@File : arrangeCoins
@Software: GoLand
*/
package arrangeCoins

func arrangeCoins(n int) int {
	// return int(math.Sqrt(float64(n*8) + 1) / 2 - 0.5)
	if n == 0 {
		return 0
	}
	left, right := 1, n
	for left < right {
		mid := (left + right) >> 1
		if (mid*(mid+1))>>1 < n {
			left = mid + 1
		} else {
			right = mid
		}
	}
	if left*(left+1)>>1 == n {
		return left
	}
	return left - 1
}